# Laboratorio 2 - Parte 2

### Modelos no paramétricos

### 2019-I

#### Profesor: Julián D. Arias Londoño
#### julian.ariasl@udea.edu.co


## Guía del laboratorio

En esta archivo va a encontrar tanto celdas de código cómo celdas de texto con las instrucciones para desarrollar el laboratorio.

Lea atentamente las instrucciones entregadas en las celdas de texto correspondientes y proceda con la solución de las preguntas planteadas.

Nota: no olvide ir ejecutando las celdas de código de arriba hacia abajo para que no tenga errores de importación de librerías o por falta de definición de variables.

#### Primer integrante
Nombre: Daniel Torres González
Cédula: 1020492000

#### Segundo integrante
Nombre: Daniel Santa Rendón
Cédula: 1040050021

In [1]:
from __future__ import division
%matplotlib inline
import numpy as np
import scipy as sc
import matplotlib.pyplot as plt


#Evitar algunas advertencias
import warnings
warnings.filterwarnings("always")

## Ejercicio 1: Contextualización del problema


Usaremos el dataset iris para el problema de clasificación, desde la librería sklearn. En el UCI Machine Learning Repository se encuentra más información en el siguiente [link](https://archive.ics.uci.edu/ml/datasets/iris) .

In [2]:
#cargamos la bd iris desde el dataset de sklearn
from sklearn import datasets

iris = datasets.load_iris()


X, Y = iris.data, iris.target

C:\Users\ALEJO\Anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\ALEJO\Anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\ALEJO\Anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\ALEJO\Anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\ALEJO\Anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 19

Responda las siguientes preguntas: 

1.1 ¿Cu&aacute;ntas clases tiene el problema?:


In [3]:
len(np.unique(Y))

3

In [4]:
np.unique(Y)

array([0, 1, 2])

El problema tiene 3 clases: 0, 1 y 2

1.2 ¿Cu&aacute;ntas caracter&iacute;sticas tiene el problema?:

In [5]:
X.shape[1]

4

El problema tiene 4 características

1.3 ¿Cu&aacute;ntas muestras tiene el problema?:

In [6]:
X.shape[0]

150

El problema tiene 150 muestras

1.4 ¿Cu&aacute;ntas muestras por cada clase hay en la base de datos?:

In [7]:
print("Para la clase 0 son ",len(Y[Y==0]), " muestras")
print("Para la clase 1 son ",len(Y[Y==1]), " muestras")
print("Para la clase 2 son ",len(Y[Y==2]), " muestras")

Para la clase 0 son  50  muestras
Para la clase 1 son  50  muestras
Para la clase 2 son  50  muestras


## Ejercicio 2: Completar código KNN

Analice los siguientes métodos de la teoría vista para los modelos de *K-vecinos más cercanos (KNN)* y complete el código del método KNN.

<b>Nota</b>: Para el cáculo de la distancia entre vectores tienen dos opciones, usar la función la distancia euclidiana `scipy.spatial.distance.euclidean`([Ejemplo](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.euclidean.html)) o usar la función `numpy.linalg.norm`([Ejemplo](https://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.norm.html)). Revisen la documentación para comprender su uso. También serán de utilidad las funciones `numpy.sort` y `numpy.argsort`.


<b>Nota</b>: Para el cáculo de la distancia entre vectores tienen dos opciones, usar la función scipy.spatial.distance.euclidean o usar la función numpy.linalg.norm. <b>Revisen la documentación </b> para comprender su uso. También serán de utilidad las funciones numpy.sort y numpy.argsort.

In [8]:
from scipy.stats import mode
import operator

#Para calcular el error en los problemas de clasificación
def ErrorClas(Y_lest, Y):
    error = 1 - np.sum(Y_lest == Y)/len(Y)
    
    return error


def KNN(X_train, Y_train, X_val, k, tipo):
    
    #X_train es la matriz con las muestras de entrenamiento
    #Y_train es un vector con los valores de salida pra cada una de las muestras de entrenamiento
    #X_val es la matriz con las muestras de validación
    #tipo es una bandera que indica si el problema es de regresión o de clasificación.
    
    #Parámetro k que equivale al número de vecinos a tener en cuenta para resolver el problema de predicción de la variable de salida
    Nt = len(X_val)
    Y_test = np.zeros(Nt)

    num = 0
    distances = sc.spatial.distance_matrix(X_val, X_train)
    
    for xtest in X_val:
        neighbors_index = np.argsort(distances[num])[:k]
        k_nearest = np.take(Y_train, neighbors_index)
        
        if tipo == 'r':
            Y_test[num] = np.sum(k_nearest)/k
            
        elif tipo == 'c':
            Y_test[num] = sc.stats.mode(k_nearest)[0]
            
        num += 1
    
    return Y_test #Debe retornar un vector que contenga las predicciones para cada una de las muestras en X_val, en el mismo orden.  
  

## Ejercicio 3

Una vez haya completado el codigo del método de KNN, ejecute varias veces el proceso de entrenamiento y evaluación cambiando el parametro $k$, el cual es el numero de vecinos, y complete la siguiente tabla con los valores del error de clasificación obtenidos:

In [24]:
from numpy import random
from sklearn.model_selection import train_test_split
import math
import numpy.matlib as matlib
N = np.size(X,0)

#Validamos el modelo
random.seed(1)
Error = np.zeros(4)
for j in range(4):
    
    Xtrain, Xtest, Ytrain, Ytest = train_test_split(X, Y, test_size=0.25)
    #Normalizamos los datos
    media = np.mean(Xtrain)
    desvia = np.std(Xtrain)
    Xtrain = sc.stats.stats.zscore(Xtrain)
    Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)

    Yest = KNN(Xtrain, Ytrain, Xtest, 2, 'c')#Complete con el llamado apropiado de la función que Uds diseñaron
    
    #Evaluamos las predicciones del modelo con los datos de test
    Error[j] = ErrorClas(Yest,Ytest)
print('\nError durante la prueba = ' + str(np.mean(Error)) + '+-' + str(np.std(Error)))


Error durante la prueba = 0.26973684210526316+-0.08802031684381349


Responda:

3.1 ¿Qué metodología de validación se usa en el experimento?


In [10]:
Xtrain, Xtest, Ytrain, Ytest = train_test_split(X, Y, test_size=0.25)
print(Xtrain.shape, Xtest.shape)

(112, 4) (38, 4)


El método utilizado es el de **Validación Booststrapping**, debido a que anteriormente se realizó una división aleatoria 75-25, donde el **25%** es usado para la validación y luego se repite nuevamente el proceso. 

3.2 ¿Cuántas muestras se usan para el entrenamiento?


In [11]:
Xtrain.shape[0]

112

Se utilizan 112 muestras para el entrenamiento


3.3 ¿Cuántas muestras se usan para la validación?

In [12]:
Xtest.shape[0]

38

Se utilizan 38 muestras para la validación

#### Tabla de resultados

In [13]:
def prueba (k, tipo = 'c'):
    
    N = np.size(X,0)

    #Validamos el modelo
    random.seed(1)
    Error = np.zeros(4)
    
    for j in range(4):

        Xtrain, Xtest, Ytrain, Ytest = train_test_split(X, Y, test_size=0.25)
        
        #Normalizamos los datos
        media = np.mean(Xtrain)
        desvia = np.std(Xtrain)
        Xtrain = sc.stats.stats.zscore(Xtrain)
        Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)

        Yest = KNN(Xtrain, Ytrain, Xtest, k, tipo)#Complete con el llamado apropiado de la función que Uds diseñaron

        #Evaluamos las predicciones del modelo con los datos de test
        Error[j] = ErrorClas(Yest,Ytest)
        
    return np.mean(Error), np.std(Error)

In [25]:
import pandas as pd
import qgrid
pd.options.mode.chained_assignment = None

randn = np.random.randn
df_types = pd.DataFrame({
    'Numero de vecinos' : pd.Series(['1', '2', '3', '4', '5', '6', '7'])})
df_types["Error_Prueba"] = ""
df_types["Desviación estándar del error"] = ""
df_types.set_index(['Numero de vecinos'], inplace=True)
#df_types["Error_Prueba"][1] = "0.2697"
#df_types["Desviación estándar del error"][1] = "0.088" 

i = 0
for k in df_types.index:
    error, std = prueba(int(k))
    
    df_types["Error_Prueba"][i] = str(error)
    df_types["Desviación estándar del error"][i] = str(std)
    
    i += 1
    
qgrid_widget = qgrid.show_grid(df_types, show_toolbar=False)
qgrid_widget

QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': True, 'defau…

Ejecute la siguiente instrucción para dejar guardados en el notebook los resultados de las pruebas.

In [15]:
qgrid_widget.get_changed_df()

Error_Prueba Desviación estándar del error
Numero de vecinos                                                   
1                  0.20394736842105263           0.11376063464335907
2                  0.26973684210526316           0.08802031684381349
3                  0.26973684210526316           0.08603090020146066
4                  0.28289473684210525           0.07529949435697103
5                  0.26973684210526316           0.08802031684381349
6                   0.2763157894736842            0.0842516347030638
7                   0.2763157894736842            0.0842516347030638

## Ejercicio 4

A continuación debe completar el código del método parzen para resolver problemas de clasificación con el modelo ventana de Parzen.

In [16]:
#Calcula el kernel gaussiano de x
def kernel_gaussiano(x):
    return np.exp((-0.5)*x**2)

def ParzenWindow(x,Data,h):
    h = h
    Ns = Data.shape[0]
    suma = 0
    for k in range(Ns):
        u = sc.spatial.distance.euclidean(x,Data[k,:])
        suma += kernel_gaussiano(u/h)
    return suma

def ParzenClass(X_train, Y_train, X_val, ancho_h):

    h = ancho_h
    Nv = len(X_val)
    Y_est = np.array([])  
    X_clases = []
    
    for clase in np.unique(Y_train):
        X_clases.append(X_train[np.where(Y_train == clase )])    

    probabilidades = []
    j = 0
    for x_clase in X_clases: 
        distances = sc.spatial.distance_matrix(X_val, x_clase) / h
        
        for i in range(Nv):
            k = kernel_gaussiano(distances[i])
            P = np.sum(k)
            if(j == 0):
                probabilidades.append([P, 0.0, 0.0])
            else:
                probabilidades[i][j] = P
        j += 1   
    
    for p in probabilidades:
        Y_est = np.append(Y_est,p.index(max(p)))
    return Y_est#Debe retornar un vector que contenga las predicciones para cada una de las muestras en X_val, en el mismo orden.  

## Ejercicio 5

Ahora debe resolver el mismo problema de clasificación con el método ventana de Parzen. Complete el código ejecute las pruebas para diferentes valores de $h$ y llene la siguiente tabla:

In [17]:
from numpy import random
from sklearn.model_selection import StratifiedKFold
import math

#Validamos el modelo
random.seed(1)
Error = np.zeros(4)
skf = StratifiedKFold(n_splits=4)
j = 0
MTrain = []
MTest = []

for train, test in skf.split(X, Y):
    Xtrain = X[train,:]
    MTrain.append(Xtrain.shape[0])
    Ytrain = Y[train]
    Xtest = X[test,:]
    MTest.append(Xtest.shape[0])
    Ytest = Y[test]
    #Normalizamos los datos
    media = np.mean(Xtrain)
    desvia = np.std(Xtrain)
    Xtrain = sc.stats.stats.zscore(Xtrain)
    Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)

    Yest = ParzenClass(Xtrain, Ytrain, Xtest, 0.1)#Complete con el llamado apropiado de la función que Uds diseñaron
  
    #Evaluamos las predicciones del modelo con los datos de test
    Error[j] = ErrorClas(Yest,Ytest)
    j += 1
print('\nError durante la prueba = ' + str(np.mean(Error)) + '+-' + str(np.std(Error)))


Error durante la prueba = 0.2216880341880342+-0.04332854434165311


Responda:
    
5.1 ¿Qué metodología de validación se usó en la simulación?:


In [18]:
skf = StratifiedKFold(n_splits=4)
skf.split(X, Y)

<generator object _BaseKFold.split at 0x0000028827EC6E58>

El metodo que se utiliza en este experimentos es el de **Validación Cruzada K folds** usando 4 folks, en la linea anterior que hace parte del codigo lo que realiza es dividir las muestras en 4 grupos distintos en el que se utilizaran 3 para el entrenamiento y 1 para la validacion. 

5.2 ¿Cuántas muestras se usan para el entrenamiento?:

In [19]:
print(MTrain)

[111, 111, 114, 114]


Como en el método se hizo con 4 folds, por lo tanto, fueron 4 iteraciones; para cada iteración es una cantidad diferente: Para la primera iteración fueron 111, igualmente que para la segunda. Para la tercera y la cuarta iteración fueron 114 cada una

5.3 ¿Cuántas muestras se usan para la validación?:

In [20]:
print(MTest)

[39, 39, 36, 36]


Para cada fold, también fue una cantidad de muestras difernte en el entrenamiento. Para la primera y la segunda iteración se usaron 39 muetsras. Para las dos iteraciones restantes se usaron 36 muestras en cada una.

#### Tabla de resultados

In [21]:
def pruebaVP(h):
    
    random.seed(1)
    Error = np.zeros(4)
    skf = StratifiedKFold(n_splits=4)
    j = 0
    for train, test in skf.split(X, Y):
        Xtrain = X[train,:]
        Ytrain = Y[train]
        Xtest = X[test,:]
        Ytest = Y[test]
        #Normalizamos los datos
        media = np.mean(Xtrain)
        desvia = np.std(Xtrain)
        Xtrain = sc.stats.stats.zscore(Xtrain)
        Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)

        Yest = ParzenClass(Xtrain, Ytrain, Xtest, h)#Complete con el llamado apropiado de la función que Uds diseñaron

        #Evaluamos las predicciones del modelo con los datos de test
        Error[j] = ErrorClas(Yest,Ytest)
        j += 1
    
    return np.mean(Error), np.std(Error)

In [22]:
import pandas as pd
import qgrid
randn = np.random.randn
df_types2 = pd.DataFrame({
    'Tamano de ventana' : pd.Series(['0.05', '0.1', '0.5', '1', '2', '5', '10'])})
df_types2["Error_Prueba"] = ""
df_types2["Desviación estándar del error"] = ""
df_types2.set_index(['Tamano de ventana'], inplace=True)
df_types2["Error_Prueba"][1] = "0.222"
df_types2["Desviación estándar del error"][1] = "0.043"
#df_types.sort_index(inplace=True)

i = 0
for k in df_types2.index:
    error, std = pruebaVP(float(k))
    
    df_types2["Error_Prueba"][i] = str(error)
    df_types2["Desviación estándar del error"][i] = str(std)
    
    i += 1

qgrid_widget = qgrid.show_grid(df_types2, show_toolbar=False)
qgrid_widget

QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': True, 'defau…

Ejecute la siguiente instrucción para dejar guardados en el notebook los resultados de las pruebas.

In [23]:
qgrid_widget.get_changed_df()

Error_Prueba Desviación estándar del error
Tamano de ventana                                                   
0.05                0.2216880341880342           0.04332854434165311
0.1                 0.2216880341880342           0.04332854434165311
0.5                  0.281517094017094           0.04021339704891218
1                    0.280982905982906           0.03145818774430656
2                  0.24198717948717952          0.058466142137675314
5                   0.2216880341880342           0.04696907336459245
10                  0.2216880341880342           0.04696907336459245